front50-core/src/main/groovy/com/netflix/spinnaker/front50/listeners/ChaosMonkeyEventListener.java

```diff
@@ -0,0 +1,178 @@
+      // Flag didn't change, we don't need to do anything.
+      return updatedApplication;
+    }
+
+    Object rawPermissions = updatedApplication.details().get("permissions");
+    if (!(rawPermissions instanceof Map)) {
+      // Exit early, no permissions config found on the application
+      log.warn("No permissions config found on application '{}'", updatedApplication.getName());
+      return updatedApplication;
+    }
+
+    @SuppressWarnings("unchecked")
+    Map<String, List<String>> permissionsMap = (Map<String, List<String>>) rawPermissions;
+
+    if (isChaosMonkeyEnabled) {
+      // Chaos Monkey is enabled, apply permissions
+      List<String> readPermissions = permissionsMap.get("READ");
+      if (!readPermissions.contains(properties.getUserRole())) {
+        log.info("Chaos Monkey READ permissions applied for '{}'", updatedApplication.getName());
+        readPermissions.add(properties.getUserRole());
+      }
+      List<String> writePermissions = permissionsMap.get("WRITE");
+      if (!writePermissions.contains(properties.getUserRole())) {
+        log.info("Chaos Monkey WRITE permissions applied for '{}'", updatedApplication.getName());
+        writePermissions.add(properties.getUserRole());
+      }
+    } else {
+      // Chaos Monkey is disabled, revoke permissions
+      log.info(
+          "Revoking Chaos Monkey READ and WRITE permissions for '{}'",
+          updatedApplication.getName());
+      permissionsMap.get("READ").remove(properties.getUserRole());
+      permissionsMap.get("WRITE").remove(properties.getUserRole());
+    }
+
+    return updatedApplication;
+  }
+
+  @Override
+  public void rollback(Application originalApplication) {
+    // Do nothing.
+  }
+
+  @Override
+  public boolean supports(ApplicationPermissionEventListener.Type type) {
+    return properties.isEnabled()
+        && Arrays.asList(
+                ApplicationPermissionEventListener.Type.PRE_CREATE,
+                ApplicationPermissionEventListener.Type.PRE_UPDATE)
+            .contains(type);
+  }
+
+  @Nullable
+  @Override
+  public Application.Permission call(
+      @Nullable Application.Permission originalPermission,
+      @Nullable Application.Permission updatedPermission) {
+    if (updatedPermission == null) {
+      return null;
+    }
+
+    Application application = applicationDAO.findByName(updatedPermission.getName());
+
+    if (isChaosMonkeyEnabled(application)) {
+      applyNewPermissions(updatedPermission, true);
+    } else {
+      applyNewPermissions(updatedPermission, false);
+    }
+
+    return updatedPermission;
+  }
+
+  private void applyNewPermissions(Application.Permission updatedPermission, boolean addRole) {
+    Permissions permissions = updatedPermission.getPermissions();
+
+    Map<Authorization, List<String>> data = new HashMap<>();
+    for (Authorization auth : Authorization.values()) {
+      List<String> roles = new ArrayList<>(permissions.get(auth));
```

Comment:
I notice we do this type of unpacking and rebuilding of Permissions in many places.  Maybe we should add a public method on Permissions like this:

```
  public Map<Authorization, List<String>> unpack() {
    return Arrays.stream(Authorization.values()).collect(toMap(identity(), this::get));
  }
```

So we could just call `permissions.unpack()`.

Reply:
Yea or nay? https://github.com/spinnaker/fiat/pull/532